File: packages/go/fayda/token.go
package fayda

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type cachedPublicKey struct {
	Kid string
	Key *rsa.PublicKey
}

// TokenModule handles token exchange and verification
type TokenModule struct {
	config    *Config
	http      *http.Client
	jwksMu    sync.Mutex
	jwksKeys  []cachedPublicKey
	jwksExp   time.Time
}

// Exchange exchanges an authorization code for tokens
func (t *TokenModule) Exchange(ctx context.Context, code, codeVerifier string) (*FaydaTokens, error) {
	if t.config.Sandbox {
		return &FaydaTokens{
			AccessToken: "mock-access-token-abc123def456",
			IDToken:     "mock-id-token-xyz789uvw012",
			TokenType:   "Bearer",
			ExpiresIn:   3600,
		}, nil
	}

	assertion, err := t.buildClientAssertion()
	if err != nil {
		return nil, err
	}

	form := url.Values{
		"grant_type":            {"authorization_code"},
		"client_id":             {t.config.ClientID},
		"code":                  {code},
		"redirect_uri":          {t.config.RedirectURI},
		"code_verifier":         {codeVerifier},
		"client_assertion_type": {"urn:ietf:params:oauth:client-assertion-type:jwt-bearer"},
		"client_assertion":      {assertion},
	}

	req, err := http.NewRequestWithContext(ctx, "POST", t.config.TokenEndpoint(), strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := t.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		var errResp map[string]interface{}
		json.Unmarshal(body, &errResp)
		errCode := ""
		if c, ok := errResp["error"].(string); ok {
			errCode = c
		}
		description := ""
		if d, ok := errResp["error_description"].(string); ok {
			description = d
		}
		return nil, MapTokenError(errCode, description)
	}

	var tokens FaydaTokens
	if err := json.Unmarshal(body, &tokens); err != nil {
		return nil, err
	}

	return &tokens, nil
}

// rsaJWK is a minimal JWK representation for RSA private keys
type rsaJWK struct {
	N  string `json:"n"`
	E  string `json:"e"`
	D  string `json:"d"`
	P  string `json:"p"`
	Q  string `json:"q"`
	Dp string `json:"dp"`
	Dq string `json:"dq"`
	Qi string `json:"qi"`
}

// parseRSAPrivateKey parses a JWK JSON blob into an *rsa.PrivateKey using stdlib only
func parseRSAPrivateKey(jwkBytes []byte) (*rsa.PrivateKey, error) {
	var k rsaJWK
	if err := json.Unmarshal(jwkBytes, &k); err != nil {
		return nil, fmt.Errorf("invalid JWK JSON: %w", err)
	}

	decodeBigInt := func(field, b64 string) (*big.Int, error) {
		if b64 == "" {
			return nil, fmt.Errorf("JWK missing required field: %s", field)
		}
		b, err := base64.RawURLEncoding.DecodeString(b64)
		if err != nil {
			return nil, fmt.Errorf("JWK field %s: %w", field, err)
		}
		return new(big.Int).SetBytes(b), nil
	}

	n, err := decodeBigInt("n", k.N)
	if err != nil {
		return nil, err
	}
	e, err := decodeBigInt("e", k.E)
	if err != nil {
		return nil, err
	}
	d, err := decodeBigInt("d", k.D)
	if err != nil {
		return nil, err
	}
	p, err := decodeBigInt("p", k.P)
	if err != nil {
		return nil, err
	}
	q, err := decodeBigInt("q", k.Q)
	if err != nil {
		return nil, err
	}
	dp, err := decodeBigInt("dp", k.Dp)
	if err != nil {
		return nil, err
	}
	dq, err := decodeBigInt("dq", k.Dq)
	if err != nil {
		return nil, err
	}
	qi, err := decodeBigInt("qi", k.Qi)
	if err != nil {
		return nil, err
	}

	key := &rsa.PrivateKey{
		PublicKey: rsa.PublicKey{
			N: n,
			E: int(e.Int64()),
		},
		D:      d,
		Primes: []*big.Int{p, q},
		Precomputed: rsa.PrecomputedValues{
			Dp:   dp,
			Dq:   dq,
			Qinv: qi,
		},
	}

	if err := key.Validate(); err != nil {
		return nil, fmt.Errorf("RSA key validation failed: %w", err)
	}

	return key, nil
}

// buildClientAssertion builds the RS256-signed JWT client assertion
func (t *TokenModule) buildClientAssertion() (string, error) {
	// Decode base64-encoded JWK string
	jwkBytes, err := base64.StdEncoding.DecodeString(t.config.PrivateKeyBase64)
	if err != nil {
		return "", fmt.Errorf("failed to decode private key base64: %w", err)
	}

	rsaKey, err := parseRSAPrivateKey(jwkBytes)
	if err != nil {
		return "", fmt.Errorf("failed to parse JWK: %w", err)
	}

	now := time.Now()
	claims := jwt.MapClaims{
		"iss": t.config.ClientID,
		"sub": t.config.ClientID,
		"aud": t.config.TokenEndpoint(),
		"iat": now.Unix(),
		"exp": now.Add(2 * time.Hour).Unix(),
		"jti": generateUUID(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	return token.SignedString(rsaKey)
}

// fetchJWKS fetches and caches Fayda's public JWKS (1-hour TTL)
func (t *TokenModule) fetchJWKS(ctx context.Context) ([]cachedPublicKey, error) {
	t.jwksMu.Lock()
	defer t.jwksMu.Unlock()

	if !t.jwksExp.IsZero() && time.Now().Before(t.jwksExp) {
		return t.jwksKeys, nil
	}

	// 1. Discovery document → jwks_uri
	req, _ := http.NewRequestWithContext(ctx, "GET", t.config.DiscoveryEndpoint(), nil)
	resp, err := t.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("discovery fetch failed: %w", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var discovery map[string]interface{}
	if err := json.Unmarshal(body, &discovery); err != nil {
		return nil, fmt.Errorf("discovery parse failed: %w", err)
	}
	jwksURI, _ := discovery["jwks_uri"].(string)
	if jwksURI == "" {
		return nil, fmt.Errorf("jwks_uri not found in discovery document")
	}

	// 2. Fetch JWKS
	req2, _ := http.NewRequestWithContext(ctx, "GET", jwksURI, nil)
	resp2, err := t.http.Do(req2)
	if err != nil {
		return nil, fmt.Errorf("JWKS fetch failed: %w", err)
	}
	defer resp2.Body.Close()
	body2, _ := io.ReadAll(resp2.Body)
	var jwks struct {
		Keys []map[string]interface{} `json:"keys"`
	}
	if err := json.Unmarshal(body2, &jwks); err != nil {
		return nil, fmt.Errorf("JWKS parse failed: %w", err)
	}

	// 3. Parse RSA public keys
	var keys []cachedPublicKey
	for _, k := range jwks.Keys {
		if k["kty"] != "RSA" {
			continue
		}
		nStr, _ := k["n"].(string)
		eStr, _ := k["e"].(string)
		nBytes, err := base64.RawURLEncoding.DecodeString(nStr)
		if err != nil {
			continue
		}
		eBytes, err := base64.RawURLEncoding.DecodeString(eStr)
		if err != nil {
			continue
		}
		kid, _ := k["kid"].(string)
		keys = append(keys, cachedPublicKey{
			Kid: kid,
			Key: &rsa.PublicKey{
				N: new(big.Int).SetBytes(nBytes),
				E: int(new(big.Int).SetBytes(eBytes).Int64()),
			},
		})
	}

	t.jwksKeys = keys
	t.jwksExp = time.Now().Add(time.Hour)
	return keys, nil
}

// Verify verifies an ID token's RS256 signature against Fayda's public JWKS.
// Returns the decoded MapClaims on success.
func (t *TokenModule) Verify(ctx context.Context, idToken string) (jwt.MapClaims, error) {
	keys, err := t.fetchJWKS(ctx)
	if err != nil {
		return nil, err
	}
	var lastErr error
	for _, k := range keys {
		entry := k
		token, err := jwt.Parse(idToken, func(tok *jwt.Token) (interface{}, error) {
			if _, ok := tok.Method.(*jwt.SigningMethodRSA); !ok {
				return nil, fmt.Errorf("unexpected signing method: %v", tok.Header["alg"])
			}
			if kid, ok := tok.Header["kid"].(string); ok && entry.Kid != "" && kid != entry.Kid {
				return nil, fmt.Errorf("kid mismatch")
			}
			return entry.Key, nil
		}, jwt.WithAudience(t.config.ClientID))
		if err == nil && token.Valid {
			return token.Claims.(jwt.MapClaims), nil
		}
		lastErr = err
	}
	if lastErr != nil {
		return nil, fmt.Errorf("ID token verification failed: %w", lastErr)
	}
	return nil, fmt.Errorf("no valid JWKS key found for token")
}

// generateUUID returns a random RFC 4122 version 4 UUID
func generateUUID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		panic(fmt.Sprintf("crypto/rand failed: %v", err))
	}
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant bits
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}

File: packages/rust/src/token.rs
use crate::{config::FaydaConfig, errors::FaydaError, models::FaydaTokens};
use base64::Engine;
use jsonwebtoken::{
    decode, decode_header, encode, Algorithm, DecodingKey, EncodingKey, Header, Validation,
};
use rsa::{pkcs8::EncodePrivateKey, BigUint, RsaPrivateKey};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

pub struct Token {
    config: FaydaConfig,
}

#[derive(Serialize, Deserialize)]
struct ClientAssertionClaims {
    iss: String,
    sub: String,
    aud: String,
    iat: u64,
    exp: u64,
    jti: String,
}

impl Token {
    pub fn new(config: FaydaConfig) -> Self {
        Token { config }
    }

    fn build_client_assertion(&self) -> Result<String, FaydaError> {
        use base64::engine::general_purpose::{STANDARD, URL_SAFE_NO_PAD};

        // Decode base64-encoded JWK string
        let jwk_bytes = STANDARD
            .decode(&self.config.private_key_base64)
            .map_err(|e| FaydaError::Config(format!("base64 decode error: {}", e)))?;

        let jwk: serde_json::Value = serde_json::from_slice(&jwk_bytes)?;

        // Decode a base64url JWK field to BigUint
        let decode_field = |field: &str| -> Result<BigUint, FaydaError> {
            let b64 = jwk[field]
                .as_str()
                .ok_or_else(|| FaydaError::Config(format!("JWK missing field: {}", field)))?;
            let bytes = URL_SAFE_NO_PAD
                .decode(b64)
                .map_err(|e| FaydaError::Config(format!("base64 decode '{}': {}", field, e)))?;
            Ok(BigUint::from_bytes_be(&bytes))
        };

        let n = decode_field("n")?;
        let e_val = decode_field("e")?;
        let d = decode_field("d")?;
        let p = decode_field("p")?;
        let q = decode_field("q")?;

        let rsa_key = RsaPrivateKey::from_components(n, e_val, d, vec![p, q])
            .map_err(|e| FaydaError::Config(format!("RSA key construction error: {}", e)))?;

        let pem = rsa_key
            .to_pkcs8_pem(rsa::pkcs8::LineEnding::LF)
            .map_err(|e| FaydaError::Config(format!("PEM encoding error: {}", e)))?;

        let encoding_key = EncodingKey::from_rsa_pem(pem.as_bytes())
            .map_err(|e| FaydaError::Config(format!("EncodingKey error: {}", e)))?;

        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(|e| FaydaError::Config(format!("system time error: {}", e)))?
            .as_secs();

        let claims = ClientAssertionClaims {
            iss: self.config.client_id.clone(),
            sub: self.config.client_id.clone(),
            aud: self.config.token_endpoint(),
            iat: now,
            exp: now + 7200,
            jti: Uuid::new_v4().to_string(),
        };

        let header = Header::new(Algorithm::RS256);
        encode(&header, &claims, &encoding_key).map_err(FaydaError::Jwt)
    }

    /// Exchange authorization code for tokens
    pub async fn exchange(
        &self,
        code: &str,
        code_verifier: &str,
    ) -> Result<FaydaTokens, FaydaError> {
        if self.config.sandbox {
            return Ok(FaydaTokens {
                access_token: "mock-access-token-abc123def456".to_string(),
                id_token: "mock-id-token-xyz789uvw012".to_string(),
                token_type: "Bearer".to_string(),
                expires_in: 3600,
            });
        }

        let assertion = self.build_client_assertion()?;

        let params = [
            ("grant_type", "authorization_code"),
            ("client_id", self.config.client_id.as_str()),
            ("code", code),
            ("redirect_uri", self.config.redirect_uri.as_str()),
            ("code_verifier", code_verifier),
            (
                "client_assertion_type",
                "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            ),
            ("client_assertion", assertion.as_str()),
        ];

        let client = reqwest::Client::new();
        let response = client
            .post(self.config.token_endpoint())
            .form(&params)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_body: serde_json::Value = response.json().await?;
            let error_code = error_body["error"].as_str().unwrap_or("").to_string();
            let description = error_body["error_description"]
                .as_str()
                .unwrap_or("")
                .to_string();
            return Err(crate::errors::map_token_error(&error_code, &description));
        }

        let tokens: FaydaTokens = response.json().await?;
        Ok(tokens)
    }

    /// Verify an ID token's RS256 signature against Fayda's public JWKS.
    /// Returns the decoded claims as a JSON map on success.
    pub async fn verify(&self, id_token: &str) -> Result<HashMap<String, Value>, FaydaError> {
        // Extract kid from token header (optional — used to select the right key)
        let kid = decode_header(id_token).ok().and_then(|h| h.kid);

        let http = reqwest::Client::new();

        // Fetch discovery document → jwks_uri
        let discovery: Value = http
            .get(self.config.discovery_endpoint())
            .send()
            .await?
            .json()
            .await?;

        let jwks_uri = discovery["jwks_uri"]
            .as_str()
            .ok_or_else(|| FaydaError::Config("jwks_uri not found in discovery document".into()))?;

        // Fetch JWKS
        let jwks: Value = http.get(jwks_uri).send().await?.json().await?;

        let keys = jwks["keys"]
            .as_array()
            .ok_or_else(|| FaydaError::Config("JWKS missing keys array".into()))?;

        // Find the matching key (by kid if present, otherwise first RSA key)
        let key_json = keys
            .iter()
            .find(|k| {
                if let Some(ref kid_val) = kid {
                    k["kid"].as_str() == Some(kid_val.as_str())
                } else {
                    k["kty"].as_str() == Some("RSA")
                }
            })
            .ok_or_else(|| FaydaError::Config("No matching RSA key found in JWKS".into()))?;

        let n = key_json["n"]
            .as_str()
            .ok_or_else(|| FaydaError::Config("JWK missing 'n'".into()))?;
        let e = key_json["e"]
            .as_str()
            .ok_or_else(|| FaydaError::Config("JWK missing 'e'".into()))?;

        let decoding_key = DecodingKey::from_rsa_components(n, e).map_err(FaydaError::Jwt)?;

        let mut validation = Validation::new(Algorithm::RS256);
        validation.set_audience(&[self.config.client_id.as_str()]);

        let token_data = decode::<HashMap<String, Value>>(id_token, &decoding_key, &validation)
            .map_err(FaydaError::Jwt)?;

        Ok(token_data.claims)
    }
}
